internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,360 @@
+	allErrs = append(allErrs,
+		validatePatchDefinitions(patch, clusterClass, path)...,
+	)
+	return allErrs
+}
+
+func validatePatchName(patch clusterv1.ClusterClassPatch, names sets.String, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if patch.Name == "" {
+		allErrs = append(allErrs,
+			field.Required(
+				path.Child("name"),
+				"patch name must be defined",
+			),
+		)
+	}
+
+	if names.Has(patch.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				fmt.Sprintf("patch names must be unique. Patch with name %q is defined more than once.", patch.Name),
+			),
+		)
+	}
+	return allErrs
+}
+
+func validatePatchDefinitions(patch clusterv1.ClusterClassPatch, clusterClass *clusterv1.ClusterClass, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for i, definition := range patch.Definitions {
+		allErrs = append(allErrs,
+			validateJSONPatches(definition.JSONPatches, clusterClass.Spec.Variables, path.Child("definitions").Index(i).Child("jsonPatches"))...)
+		allErrs = append(allErrs,
+			validateSelectors(definition.Selector, clusterClass, path.Child("definitions").Index(i).Child("selector"))...)
+	}
+	return allErrs
+}
+
+// validateSelectors tests to see if the selector matches any template in the ClusterClass.
+// It returns nil as soon as it finds any matching template and an error if there is no match.
+func validateSelectors(selector clusterv1.PatchSelector, class *clusterv1.ClusterClass, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if selector.MatchResources.InfrastructureCluster == nil && selector.MatchResources.ControlPlane == nil &&
+		(selector.MatchResources.MachineDeploymentClass == nil || len(selector.MatchResources.MachineDeploymentClass.Names) == 0) {
+		return append(allErrs,
+			field.Invalid(
+				path,
+				selector,
+				"no selector enabled",
+			))
+	}
+	if selector.MatchResources.InfrastructureCluster != nil && class.Spec.Infrastructure.Ref != nil &&
+		*selector.MatchResources.InfrastructureCluster {
+		if selectorMatchTemplate(selector, class.Spec.Infrastructure.Ref) {
+			return nil
+		}
+	}
+	if selector.MatchResources.ControlPlane != nil && class.Spec.ControlPlane.Ref != nil &&
+		*selector.MatchResources.ControlPlane {
+		if selectorMatchTemplate(selector, class.Spec.ControlPlane.Ref) {
+			return nil
+		}
+	}
+
+	if class.Spec.ControlPlane.MachineInfrastructure != nil && class.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		if selectorMatchTemplate(selector, class.Spec.ControlPlane.MachineInfrastructure.Ref) {
+			return nil
+		}
+	}
+	if selector.MatchResources.MachineDeploymentClass != nil &&
+		len(selector.MatchResources.MachineDeploymentClass.Names) > 0 {
+		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+			for _, md := range class.Spec.Workers.MachineDeployments {
+				if md.Class == name {
+					if selectorMatchTemplate(selector, md.Template.Infrastructure.Ref) {
```

Comment:
:/ Those `.Ref's` can be nil too. I think we have to nil check to avoid potential panics. 

Up to you to add the nil check here or inside of `selectorMatchTemplate`